nixos/modules/services/networking/dnsmasq.nix

```diff
@@ -19,6 +19,22 @@ let
     ${cfg.extraConfig}
   '';
 
+  /*
+    The sed command removes stateful parts of the config that won't work in this
+    derivation.  This method should be fine, as according to dnsmasq(8) (section
+    CONFIG FILE) "the format of this file consist of one option per line". It
+    does not mention any exceptions to this rule. The only caveat is that this
+    will not catch any problems with the config file involving options that
+    contain the string /etc/, but it's better than falsely rejecting a valid
+    config file.
+  */
+  configChecked = pkgs.runCommand "dnsmasq-config-checked" {} ''
+    sed '/\/etc\//d' ${dnsmasqConf} > dnsmasq.conf
+
+    ${dnsmasq}/bin/dnsmasq --test -C dnsmasq.conf && touch $out
+    echo "dnsmasq config file: ${dnsmasqConf}"
```

Comment:
Are you sure, this will not break in sandbox builds? http://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=blob;f=src/option.c;h=1a5e359b8cabd045e25db082444035649871bcc4;hb=HEAD#l4842

I see `gethostname` here for example.

Reply:
`gethostname` should just return "localhost" in a sandbox build.

Reply:
That said, I'm not sure there aren't other ways this could end up failing.

Reply:
What about files that are not in `/etc/`, nixops for example put secrets in `/run/keys/`.
Also I am in favor for having this feature I do not see a way, how this would work in sandbox mode properly.

Reply:
Yeah… What about adding an option to opt-in or opt-out of this check, and still test the configuration on startup?

Reply:
Maybe. If we would allow impure nixos evaluation there are modules that could be tested. nginx for example.

Reply:
Even that wouldn't be enough, breaking for instance any references to `/etc` that aren't valid in the current system but are in the system we're building. Best to just leave the testing for activation time.